Generate random ID for postcode regions when syncing shipping settings

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

To sync the shipping rates for postal codes of a country, we need to define them as "[regions](https://developers.google.com/shopping-content/reference/rest/v2.1/regions#Region)" through the Google API. Even though we do not need to do this explicitly, Google automatically creates "[regions](https://developers.google.com/shopping-content/reference/rest/v2.1/regions#Region)" when we present the postal codes as part of the [ShippingSettings API under `postalCodeGroups`](https://developers.google.com/shopping-content/guides/shippingsettings#set_postal_code_groups). 

The ["name" property of PostalCodeGroup](https://developers.google.com/shopping-content/reference/rest/v2.1/shippingsettings#PostalCodeGroup) does not have any explicit requirements defined for it. However, since Google maps PostalCodeGroups to Regions, the [PostalCodeGroup's](https://developers.google.com/shopping-content/reference/rest/v2.1/shippingsettings#PostalCodeGroup) `name` parameter maps directly to a [Region's](https://developers.google.com/shopping-content/reference/rest/v2.1/regions#Region) `regionId` parameter.

The `regionId` parameter has some [requirements defined by Google](https://support.google.com/merchants/answer/9698880?hl=en#requirements). One of the main requirements is the following:

> For privacy reasons, the region ID value must be a randomized set of numbers (minimum 6 digits), rather than a readable name, such as "NY_City". Readable names or names that look like postal codes aren't accepted.

Previously, we added the country and state information in all region IDs. However, in this PR, the `regionId` parameter is set to a randomly generated number containing >= 6 digits.

Note that this random ID is generated on the fly whenever the user syncs their shipping settings and we do not store this ID in the local DB. This means that whenever the shipping settings are updated, a new random ID will be generated for each region. 

The above won't affect our current functionality because we do not use this ID to identify a region in our code. However, another [requirement](https://support.google.com/merchants/answer/9698880?hl=en#requirements) is:

> You’re able to accept the region ID parameter passed by Google into your product landing pages and render the correct regional product pages with corresponding regional pricing and availability.

This functionality is not implemented right now and I don't believe we have any plans for it. Moreover, all shipping data is stored on the Google side and we can retrieve them from the [`regions` API](https://developers.google.com/shopping-content/reference/rest/v2.1/regions) if we ever need to identify the regions.


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Make sure you have a shipping zone for the "US" country that includes zip codes and a flat rate method.
2. Edit your GLA free listings settings and choose the "Recommended" settings for the shipping option.
3. Save the free listing so that the shipping settings are synced to MC.
4. Open your MC and go to the shipping & return settings.
5. Check the details of the shipping service that was just synced. It will have a name similar to "[63a5] Google Listings and Ads generated service - US"
6. Go to the "Cost" section and click on "All products"
7. Under the "Destination" column of "Shipping rate table" there should be a random number with more than 6 digits.
8. Click on the pencil mark (Edit) right next to that random number to open the "Edit destination" popup.
9. Confirm that the postcodes you entered in your shipping zone all appear under "Postal codes" in that popup.  


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Tweak - Generate random ID for postcode regions when syncing shipping settings

Commit:
Create ShippingRegion class

Commit:
Add unit test for ShippingRegion class

Commit:
Replace postcode ranges with shipping region

Commit:
Fix unit tests

Commit:
Fix unit test

ianlin reviewed (commented):
Thanks for adding this and the good findings on

> the [PostalCodeGroup's](https://developers.google.com/shopping-content/reference/rest/v2.1/shippingsettings#PostalCodeGroup) name parameter maps directly to a [Region's](https://developers.google.com/shopping-content/reference/rest/v2.1/regions#Region) regionId parameter.

and 

> For privacy reasons, the region ID value must be a randomized set of numbers (minimum 6 digits), rather than a readable name, such as "NY_City". Readable names or names that look like postal codes aren't accepted.
> 
> Previously, we added the country and state information in all region IDs. However, in this PR, the regionId parameter is set to a randomly generated number containing >= 6 digits.

I've tested locally and it worked as expected, I can see on the MC the region ID is >= 6 digits and the postcodes are displayed correctly.

However there was an error when syncing to Google if the combination is:

Regions: `US` and `California, US`
Postcodes: `94110`

<img width="1319" alt="gla-US-CA-94110" src="https://user-images.githubusercontent.com/914406/175916276-8b798ade-f045-4275-bb03-8069109207d1.png">

![gla-US-CA-94110-error](https://user-images.githubusercontent.com/914406/175916336-ee815263-670a-4153-93f8-d201616d1f9b.png)

I understand that one can only choose `Destination by postal code` or `Destination by state` from the MC but not both, and this combination may not be common in the real world. Then I tested some different combinations and the results are shown below:

## Two Regions without Postcodes

<img width="1327" alt="gla-US-CA" src="https://user-images.githubusercontent.com/914406/175916956-ce6c0ff5-5781-40b2-877e-f92f0b2fee66.png">

It will create two records in MC:

<img width="1107" alt="gla-US-CA-MC" src="https://user-images.githubusercontent.com/914406/175917038-980b06c5-cbdb-4864-b8be-6f8bd6c02cf6.png">

<img width="1104" alt="gla-US-CA-MC-ALL" src="https://user-images.githubusercontent.com/914406/175917110-df737a11-3b97-43d7-b582-ffc7e2231eab.png">

<img width="1116" alt="gla-US-CA-MC-CA" src="https://user-images.githubusercontent.com/914406/175917160-d3f0849d-aefa-410b-aafb-b63bd5d4af6c.png">

## One Region (`US`) with Postcodes

<img width="1323" alt="gla-US-94110" src="https://user-images.githubusercontent.com/914406/175917445-1634307d-6aff-4d67-a8b5-5493b8320f03.png">

Expected results according to this PR.

<img width="1114" alt="gla-US-94110-MC-94110" src="https://user-images.githubusercontent.com/914406/175917532-a28f7d16-a122-47c7-a34c-14cd3e2bef9d.png">

## One Region (`California, US`) with Postcodes

<img width="1328" alt="gla-CA-94110" src="https://user-images.githubusercontent.com/914406/175917679-5a4c9f06-8be0-4703-b192-f59be0d8b520.png">

Expected results according to this PR.

<img width="1075" alt="gla-CA-94110-MC-94110" src="https://user-images.githubusercontent.com/914406/175917748-144d9e28-5d9a-4078-bde4-5a43351fefd0.png">

## Using `develop` branch

All of the above combinations (including the one returns error) have the same result on MC:

<img width="1107" alt="Screenshot 2022-06-27 at 18 14 25" src="https://user-images.githubusercontent.com/914406/175919200-d1863115-f903-4c2e-b319-416843481834.png">

Commit:
Fix method return type doc

Commit:
Avoid syncing duplicate postcodes

The `__toString` method is used in the `ShippingZone::map_rates_to_locations` method to map the location rates. By adding the postcode ranges to the returned string, we essentially prevent adding duplicate rates for the same postcode ranges in the same country.

nima-karimi commented:
https://github.com/woocommerce/google-listings-and-ads/pull/1574/commits/49e4ee96499f16f9433c061b4559b815594a46fc should fix the issue with duplicate postcodes.

Ready for another review, @ianlin  🙏

ianlin reviewed (approved):
Thanks for the changes, I've tested again and the error has gone, I can see the postcodes are synced in the MC. 👍
